core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -6984,9 +6984,9 @@ void testGroupExpressionEquivalenceParams() {
 
   @Test void testCastAsCollectionType() {
     sql("select cast(array[1,null,2] as int array) from (values (1))")
-        .columnType("INTEGER NOT NULL ARRAY NOT NULL");
+        .columnType("INTEGER ARRAY NOT NULL");
```

Comment:
Could you also try to add tests for map/row/multiset types?

Reply:
add test for multiset types and row.

because calcite doesn't support map type in parser, so i doesn't add.

```
java.lang.RuntimeException: Error while parsing query: select cast(map['washington', 1, null, 44] as map[CHAR(10), INTEGER NOT NULL>] from (values (1))
	at org.apache.calcite.sql.test.AbstractSqlTester.parseAndValidate(AbstractSqlTester.java:157)
	at org.apache.calcite.sql.test.AbstractSqlTester.validateAndThen(AbstractSqlTester.java:248)
	at org.apache.calcite.sql.test.AbstractSqlTester.checkColumnType(AbstractSqlTester.java:177)
	at org.apache.calcite.test.SqlValidatorFixture.columnType(
```

Reply:
add this grammer should not support select cast(b as row(f0 int not null, f1 varchar null) , i will fix it a new pr latter

sql standard below
```
<cast specification> ::=
CAST <left paren> <cast operand> AS <cast target> <right paren>
<cast operand> ::=
<value expression>
| <implicitly typed value specification>
<cast target> ::=
<domain name>
| <data type>

<data type> ::=
<predefined type>
| <row type>
| <path-resolved user-defined type name>
| <reference type>
| <collection type>

<row type> ::=
ROW <row type body>
<row type body> ::=
<left paren> <field definition> [ { <comma> <field definition> }... ] <right paren>

<field definition> ::=
<field name> <data type>
```